### PR TITLE
[SPARK-34080][ML][PYTHON] Add UnivariateFeatureSelector

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Add UnivariateFeatureSelector

### Why are the changes needed?
Have one UnivariateFeatureSelector, so we don't need to have three Feature Selectors.

### Does this PR introduce _any_ user-facing change?
Yes
```
selector = UnivariateFeatureSelector(featureCols=["x", "y", "z"], labelCol=["target"], featureType="categorical", labelType="continuous", selectorType="numTopFeatures",  numTopFeatures=100)
```

Or

numTopFeatures
```
selector = UnivariateFeatureSelector(featureCols=["x", "y", "z"], labelCol=["target"], scoreFunction="f_classif", selectorType="numTopFeatures",  numTopFeatures=100)
```

### How was this patch tested?
Add Unit test

### CODE REVIEW

File: python/pyspark/ml/feature.pyi
Comment (HyukjinKwon) [inline python/pyspark/ml/feature.pyi:1625]: cc @zero323 FYI

File: mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala:47]: Should we change "3.1.0" to "3.1.1"?

File: mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala:181]: We have only three test functions, and there are three params to choose one, I personally think it is a over-kill.

File: mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala:246]: Is this clearer:
```
require(isSet(featureType) && isSet(labelType), "featureType and labelType need to be set")
($(featureType), $(labelType)) match {
  case ("categorical", "categorical") =>
    ChiSquareTest.test(dataset.toDF, getFeaturesCol, getLabelCol, true)
  case ("continuous", "categorical") =>
    ANOVATest.test(dataset.toDF, getFeaturesCol, getLabelCol, true)
  case ("continuous", "continuous") =>
    FValueTest.test(dataset.toDF, getFeaturesCol, getLabelCol, true)
  case _ =>
    throw new IllegalStateException(s"Unsupported combination: featureType=${$(featureType)}, labelType=${$(labelType)}")
}
```

Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala:246]: seems this exception should be `IllegalArgumentException`?

File: mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala
Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala:47]: ping @HyukjinKwon , we are trying to put this pr in 3.1. Should we use '3.1.0' here?

File: mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala:77]: Shall we make 2 params instead of one for each selection method?

* `selectionMode`: numTopFeatures, percentile, fpr, fdr, fwe
* `selectionThreshold`: double value for all

Or we can just call them "mode" and "threshold". Slightly prefer the full name, but no strong preference.

See also: https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.GenericUnivariateSelect.html#sklearn.feature_selection.GenericUnivariateSelect

File: mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala:423]: private?

File: mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala:452]: private?

File: mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala:473]: private?

**Review by mengxr — CHANGES_REQUESTED** [review]
Shall we also hide the added test classes in this release? My main concern is that they are not very practical for big data. If there are valid use cases, we should see more requests from the community.

File: mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala
Comment (HyukjinKwon) [reply mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala:47]: Oh it should be 3.1.1

File: mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala:121]: we use single `selectionThreshold` instead of original params:

- numTopFeatures, >0, 50 by default;
- percentile, [0, 1], 0.1 by default;
- fpr, [0, 1], 0.05 by default;
- fdr, [0, 1], 0.05 by default;
- fwe, [0, 1], 0.05 by default;

I suggest that we remove the default value of `selectionThreshold`:
1, if `selectionThreshold` is not set, use above orignal values, like 0.05 for fwe;
2, if `selectionThreshold` is set, check it according above ranges;

File: mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala:210]: Why not write this like
```
xxx.zipWithIndex.filter { case (pValue, index) =>
     pValue <= f * (index + 1)
}.map(_._2).max()
```

Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala:210]: this rdd maybe empty:
```
scala> sc.range(0, 0).max
java.lang.UnsupportedOperationException: empty collection
  at org.apache.spark.rdd.RDD.$anonfun$reduce$4(RDD.scala:1096)
  at scala.Option.getOrElse(Option.scala:189)
  at org.apache.spark.rdd.RDD.$anonfun$reduce$1(RDD.scala:1096)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
  at org.apache.spark.rdd.RDD.withScope(RDD.scala:388)
  at org.apache.spark.rdd.RDD.reduce(RDD.scala:1076)
  at org.apache.spark.rdd.RDD.$anonfun$max$1(RDD.scala:1511)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
  at org.apache.spark.rdd.RDD.withScope(RDD.scala:388)
  at org.apache.spark.rdd.RDD.max(RDD.scala:1511)
  ... 47 elided
```

File: mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala:189]: Let's define a function like:
```
def selectIndicesFromPValues(selectionMode: String, selectionThreshold: Double): Array[Int]
```
and then add unit test on this.
especially for the fdr case

**Review by srowen — COMMENTED** [review]
I don't mind the idea, but this seems to be removing some of the classes, like in R - is this a breaking change? esp when thinking about loading models written in previous versions. I don't think we want to do that, though we could possibly deprecate existing ones as long as they call into one new common implementation. I want to make sure this is worth the risk.

**Comment by huaxingao** [comment]
@srowen The two new feature selectors (anova classification and f regression) are not released yet. I deprecated ChiSq in this PR. I should be able to finish this PR before next RC.

File: mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala
Comment (huaxingao) [reply mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala:229]: @WeichenXu123 
I guess you want me to have more tests to verify fdr?  Is It OK with you that I add more tests for fdr and keep the methods as is? The other four selection methods are very straightforward and I guess we don't need more tests.

**Comment by HyukjinKwon** [comment]
I read some comments in the JIRA. Let me cut the RC once this one gets merged.

**Review by zhengruifeng — APPROVED** [review]
LGTM PENDING TESTS

File: mllib/src/test/scala/org/apache/spark/ml/feature/UnivariateFeatureSelectorSuite.scala
Comment (WeichenXu123) [inline mllib/src/test/scala/org/apache/spark/ml/feature/UnivariateFeatureSelectorSuite.scala:642]: Minor:
Why not directly construct a `resultDF`(only includes 2 column "pValue", "featureIndex") to test `selectIndicesFromPValues` ? This can simplify tests code.

Don't need to block merge.

**Review by WeichenXu123 — APPROVED** [review]

**Review by WeichenXu123 — APPROVED** [review]

**Comment by WeichenXu123** [comment]
Merged to apache/spark master/3.1.1
Thanks!

**Comment by huaxingao** [comment]
Thank you very much everyone!

**Pull request closed without merging**
